### PR TITLE
Add query interruption flag check to broker groupby reduction

### PR DESCRIPTION
**Pull request by jasperjiaguo**
Check for interruption during broker reduction on groupby results. So that after the timeout the worker threads can be cleaned up timely.

Screen Shot 2022-09-29 at 9 51 43 PM

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java [hunk 1]
```diff
@@ -289,6 +291,9 @@ public void runJob() {
 
               int numRows = dataTable.getNumberOfRows();
               for (int rowId = 0; rowId < numRows; rowId++) {
+                if (rowId % MAX_ROWS_UPSERT_PER_INTERRUPTION_CHECK == 0 && Thread.interrupted()) {
+                  throw new EarlyTerminationException();
```
Comment (somandal) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java:295]: nit: perhaps add an error message here?

Comment (vvivekiyer) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java:295]: Question: Is it not enough to just `return` here? Given that the main thread already threw a timeout exception? It also keeps the behavior consistent with how we return for interrupts at L279.

Comment (jasperjiaguo) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java]: Good point, changed to return nothing.

**Review by vvivekiyer — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java [hunk 2]
```diff
@@ -55,6 +55,7 @@
 import org.apache.pinot.core.transport.ServerRoutingInstance;
 import org.apache.pinot.core.util.GroupByUtils;
 import org.apache.pinot.core.util.trace.TraceRunnable;
+import org.apache.pinot.spi.exception.EarlyTerminationException;
```
Comment (vvivekiyer) [inline pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java:58]: nit: remove unused import.

File: pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java (hunk 1 above)
Comment (jasperjiaguo) [reply pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GroupByDataTableReducer.java]: Yeah as discussed below, IIUC currently no one can really catch that error. But in the future we need to add handling code here if this thread is to be canceled explicitly.

**Comment by siddharthteotia** [comment]
let's try power of 2 batch size once and assess impact of modulo v/s bit shift

**Review by siddharthteotia — APPROVED** [review]

**Pull request merged**